react/src/layout/report.tsx

```diff
@@ -351,6 +360,11 @@ export class SimulationStartLayout extends Layout<SimulationStartConfig, {}> {
         let endSimulationButton = <Button variant="primary" onClick={endSimulation}>End Simulation</Button>;
         let startSimulationButton = <Button variant="primary" onClick={startSimulation}>Start Simulation</Button>
```

Comment:
Does it make sense to make a PrimaryButton that understands how to map startSimulation to `Start Simulation`?

Reply:
Perhaps, but the definition would be `(handler, text) => <Button variant="primary" onClick={handler}>{text}</Button>` and honestly the 'End Simulation' button should probably be `variant='danger'` since it ends something irreversibly. We can always add more abstraction but it just introduces more names and more digging in cases like these. They aren't conceptually the same here. The buttons are very different and just happen to have similar definitions here but I can see them diverging pretty quickly. Soon all that will be left of the abstraction is a function that looks like `let btn = (props, text) => <Button {...props}>{text}</Button>`. Having two almost repeated lines is not enough to justify making a function because it couples these two buttons that are not what they are for the same reasons (like how ending the simulation should probably be a red button with the danger variant).

Reply:
> Perhaps, but the definition would be (handler, text) => <Button variant="primary" onClick={handler}>{text}</Button>

I was thinking of something more like this:

```jsx
(handler) => <Button variant="primary" onClick={handler}>{labelFor(handler)}</Button>
```

This has many advantages including i18n and for us consistency of names.

> HTML and JSX are long because there is a lot of information. 

Actually, there is less information in a typical HTML and JSX doc vs say YAML. (A gzip on the file will prove this.) This is why YAML is easier to read (and thusly more popular) than XML for defining state.  

Going to the code, if there was a function called `button` which accepted handler, this code could go away:

```jsx
        let endSimulationButton = <Button variant="primary" onClick={endSimulation}>End Simulation</Button>;
        let startSimulationButton = <Button variant="primary" onClick={startSimulation}>Start Simulation</Button>
```

And, the use would look like:

```jsx
                        return (
                            <Stack gap={2}>
                                <span>{'Simulation Completed'}</span>
                                <span>{`Elapsed time: ${elapsedTimeSeconds} ${'second' + (elapsedTimeSeconds !== 1 ? 's' : '')}`}</span>
                                <div>{children}</div>
                                {button(startSimulation)}
                            </Stack>
                        )
```

The abstraction is clear and contains the minimum information necessary to describe what is going to happen.

Reply:
> Going to the code, if there was a function called button which accepted handler, this code could go away:

I agree that being able to use localization would be nice but that could be handled better than coupling it with the concept of a button and handler. The buttons are also not necessarily going to be updated together. My main point was that they are not two similar lines of code because they are conceptually the same. The end simulation button should probably have a different variant defined for it anyways. After you add this as an input to the button function it would just be the same as replacing `<Button>` with `button()` and add another function developers need to know and use. It doesn't save any complexity in the long run because those two lines might not be updated together.

Reply:
> My main point was that they are not two similar lines of code because they are conceptually the same. 

Conception is in the eye of the beholder. I see them as a the same. 

Buttons in our system (not React's) have variants, handlers, and labels as I see it. A label is a concept we own, which should be centralized so that when a button (or something else) needs a label, it gets it from somewhere which manages labels. We already modularize this to a great extent. This is a basic principle of UI architecture: separating strings from formatting. It's very easy to mix the two, but it is poor separation of responsibilities to mix the two. This is why we want to avoid any strings like this in the JSX. It should all be looked up in a string table.

Buttons should have names and things like variants should be looked up as well. It's easier to reason about variants in a single table (start is primary, end danger, cancel is danger, etc.) This allows clearer grouping.

Once a button has a name, it is easy to use it to index these various tables including handlers, possibly.